core/include/vecmem/containers/impl/device_vector.ipp

```diff
@@ -225,6 +227,91 @@ VECMEM_HOST_AND_DEVICE auto device_vector<TYPE>::push_back(
     return index;
 }
 
+template <typename TYPE>
+VECMEM_HOST_AND_DEVICE auto device_vector<TYPE>::bulk_append(size_type n)
+    -> size_type {
+    // This can only be done on a resizable vector.
+    assert(m_size != nullptr);
+
+    static_assert(std::is_default_constructible<TYPE>::value,
+                  "Type `T` in `device_vector<T>::bulk_append` is not a "
+                  "default-constructible type.");
+
+    device_atomic_ref<size_type> asize(*m_size);
+    const size_type index = asize.fetch_add(n);
+    assert((index + n) <= m_capacity);
+
+    for (size_type i = 0; i < n; ++i) {
+        construct(index + i, value_type());
+    }
+
+    return index;
+}
+
+template <typename TYPE>
+VECMEM_HOST_AND_DEVICE auto device_vector<TYPE>::bulk_append(size_type n,
+                                                             const_reference v)
+    -> size_type {
+    // This can only be done on a resizable vector.
+    assert(m_size != nullptr);
+
+    device_atomic_ref<size_type> asize(*m_size);
+    const size_type index = asize.fetch_add(n);
+    assert((index + n) <= m_capacity);
+
+    for (size_type i = 0; i < n; ++i) {
+        construct(index + i, value_type(v));
+    }
+
+    return index;
+}
+
+template <typename TYPE>
+VECMEM_HOST_AND_DEVICE auto device_vector<TYPE>::bulk_append_implicit(
+    size_type n) -> size_type {
+    // This can only be done on a resizable vector.
+    assert(m_size != nullptr);
+
+#if defined(__cpp_lib_is_implicit_lifetime) && \
+    __cpp_lib_is_implicit_lifetime >= 202302L
+    static_assert(std::is_implicit_lifetyime_v<TYPE>,
+                  "Type `T` in `device_vector<T>::bulk_append_implicit` is not "
+                  "an implicit "
+                  "lifetype type (according to the C++23-compliant "
+                  "definition), so slots cannot be safely reserved.");
+#else
+    static_assert(details::is_implicit_lifetime_v<TYPE>,
+                  "Type `T` in `device_vector<T>::bulk_append_implicit` is not "
+                  "an implicit "
+                  "lifetype type (according to a non-C++23-compliant "
+                  "definition), so slots cannot be safely reserved.");
+#endif
```

Comment:
I fear these would need to be converted into SFINAE as well. I'd do it by making `vecmem::details::is_implicit_lifetime_v` "behave correctly` with all C\+\+ standards. So that the preprocessor statements would be in the declaration of that trait class, and not in its users.